nilearn/decomposition/tests/conftest.py

```diff
@@ -0,0 +1,319 @@
+            ),
+            "right": np.ones(
+                (decomposition_mesh.parts["right"].coordinates.shape[0],)
+            ),
+        }
+        return SurfaceImage(mesh=decomposition_mesh, data=mask_data)
+
+    # TODO
+    # setting the shape of the mask to be a bit different
+    # shape_3d_large that is used for the data
+    # to force resampling
+    # shape = (
+    #     shape_3d_large[0] - 1,
+    #     shape_3d_large[1] - 1,
+    #     shape_3d_large[2] - 1,
+    # )
+    shape = shape_3d_large
+    mask = np.ones(shape, dtype=np.int8)
+    mask[:5] = 0
+    mask[-5:] = 0
+    mask[:, :5] = 0
+    mask[:, -5:] = 0
+    mask[..., -2:] = 0
+    mask[..., :2] = 0
+    return Nifti1Image(mask, affine_eye)
+
+
+@pytest.fixture
+def decomposition_masker(
+    decomposition_mask_img: Union[SurfaceImage, Nifti1Image],
+    img_3d_ones_eye: Nifti1Image,
+    data_type: str,
+) -> Union[SurfaceMasker, MultiNiftiMasker]:
+    """Return the proper masker for test with volume of surface."""
+    if data_type == "surface":
+        return SurfaceMasker(mask_img=decomposition_mask_img).fit()
+    return MultiNiftiMasker(mask_img=img_3d_ones_eye).fit()
+
+
+def _decomposition_data_surface(
+    rng, decomposition_mesh, with_activation
+) -> list[SurfaceImage]:
+    surf_imgs = []
+    for _ in range(N_SUBJECTS):
+        data = {
+            "left": rng.standard_normal(
+                size=(
+                    decomposition_mesh.parts["left"].coordinates.shape[0],
+                    N_SAMPLES,
+                )
+            ),
+            "right": rng.standard_normal(
+                size=(
+                    decomposition_mesh.parts["right"].coordinates.shape[0],
+                    N_SAMPLES,
+                )
+            ),
+        }
+        if with_activation:
+            data["left"][2:4, :] += 10
+            data["right"][2:4, :] += 10
+        surf_imgs.append(SurfaceImage(mesh=decomposition_mesh, data=data))
+
+    return surf_imgs
+
+
+@pytest.fixture
+def decomposition_data(
+    decomposition_mesh,
+    rng,
+    affine_eye: np.ndarray,
+    data_type: str,
+    shape_3d_large,
+    with_activation: bool = True,
+) -> Union[list[SurfaceImage], list[Nifti1Image]]:
+    """Create "multi-subject" dataset with fake activation."""
+    if data_type == "surface":
+        return _decomposition_data_surface(
+            rng, decomposition_mesh, with_activation
+        )
+
+    nii_imgs = []
+    shape = (*shape_3d_large, N_SAMPLES)
+    for _ in range(N_SUBJECTS):
+        this_img = rng.normal(size=shape)
+        if with_activation:
+            this_img[2:4, 2:4, 2:4, :] += 10
+        nii_imgs.append(Nifti1Image(this_img, affine_eye))
+
+    return nii_imgs
+
+
+@pytest.fixture
+def decomposition_data_single_img(
+    decomposition_data,
+) -> Union[SurfaceImage, Nifti1Image]:
+    """Return a single image for decomposition."""
+    return decomposition_data[0]
```

Comment:
Could this be done other way round? I mean this could generate single image and it can be used in image generation loop in `decomposition_data` to avoid generating 5 images when you only need 1.

Reply:
let me try